### DISCUSSION TITLE
UserSession->ClientSession relationship refactoring

### DISCUSSION THREAD

**Discussion by mhajas** [keycloak/keycloak#12788] [Archive]
Co-authored by: @vramik 

# Current state
Currently, there are two separate transactions, one for the user sessions and the other for client sessions. There is one-to-many relationship between user sessions and client sessions. Each storage implementation is expected to automatically remove child entities on parent entity removal.

On the physical layer, there is a `Map<String, String>` in the user session entity that stores `client.id` and corresponding `clientSessionId`. `MapUserSessionEntity` contains the following methods:

```java
Map<String, String> getAuthenticatedClientSessions();
void setAuthenticatedClientSessions(Map<String, String> authenticatedClientSessions);
String getAuthenticatedClientSession(String clientUUID);
void setAuthenticatedClientSession(String clientUUID, String clientSessionId);
Boolean removeAuthenticatedClientSession(String clientUUID);
```

On logical layer there are separate [methods](https://github.com/keycloak/keycloak/blob/4b499c869c6279110cf65d6e69d58ae53aca82e1/server-spi/src/main/java/org/keycloak/models/UserSessionProvider.java#L42-L50) that are managing client sessions in the `UserSessionProvider`.
`MapUserSessionProvider` is implemented in a way that it first creates client session entity using `clientSessionTx.create`, then adds it to the user session via `userSessionEntity.setAuthenticatedClientSession(String clientUUID, String clientSessionId)`.

# Problem statement

In the JPA implementation to be able to model one-to-many relationship, it’s required to have access to the child entity object when it’s added to the parent collection. `JpaUserSessionEntity` doesn’t have access to the `JpaClientSessionEntity`. We’d need to introduce the entity manager into the entity itself to be able to load the object via its id. It’s not a nice solution.

On the logical layer handling of nested AuthenticatedClientSession entity is inconsistent with other similar use-cases, for example, `RootAuthenticationSession` -> `AuthenticationSession` or `RealmEntity` -> `Components`. The reason for this distinction is given by the legacy Infinispan implementation where there was a performance issue when updating one client session entity meant calling `put(UserSessionEntity)` which was causing network overhead.

# Proposed solution

## Logical layer

1. Clean logical layer - We should make the relationship between UserSessions and ClientSessions consistent which means we should aim for managing client sessions through corresponding User sessions. 
2. Client session transaction should be removed completely from the `MapUserSessionProvider`; changes to client sessions should be handled by user session transaction only. 
3. Eventually, we should remove methods managing client sessions from `MapUserSessionProvider` (may be postponed for later as legacy store still needs them)

## Physical layer

`MapUserSessionEntity` should not contain `Map<String, String>`. The methods could look something like this: 

```java
@IgnoreForEntityImplementationGenerator
void clearAuthenticatedClientSessions();
Set<MapAuthenticatedClientSessionEntity> getAuthenticatedClientSessions();
Optional<MapAuthenticatedClientSessionEntity> getAuthenticatedClientSession(String clientUUID);
void addAuthenticatedClientSession(MapAuthenticatedClientSessionEntity clientSession);
Boolean removeAuthenticatedClientSession(String clientUUID);
```
The implication of it could also be a possibility to make the `MapAuthenticatedClientSessionAdapter` class non-abstract.

## HotRod changes

The tricky part is solving the performance issue for HotRod that would be basically reintroduced by the changes mentioned above. While for JPA implementation the one-to-many relationship is handled seamlessly there is nothing like this for HotRod. The implementation of HotRod UserSession storage/transaction would need to handle this relationship manually.

Ideas for HotRod:
- HotRod User session still contain Map<String, String> mapping client.id to clientSessionId
- There are two transactions user session and client session  
- There is a Transaction aware delegate used for `UserSessionEntity` that translates the operations done on client sessions into client transaction calls. For example, if `removeAuthenticatedClientSession` is called on a user session the delegate should translate this to call on client session transaction: `clientSessionsTx.delete(clientSessionId)` etc.

**Comment by sguilhen** [comment]
Thanks for the write-up, @mhajas and @vramik!

The whole thing gets a BIG +1 from me, including the introduction of the delegate in HotRod to allow for different transactions in that particular implementation. I think all proposed changes have been carefully considered here and it will make this relationship consistent with other relationships we currently have in the map storage.

**Comment by vramik** [comment]
It might also be possible to have one config only for both user and client session. As a result we may be able to modify `MapUserSessionProviderFactory` and make it consistent with other `Map*ProviderFactory`ies.

**Reply by mhajas** [reply]
Sure, if we remove client session transaction from the provider it no longer makes sense to have two options.
